pilot/pkg/model/validation.go

```diff
@@ -1911,27 +1911,34 @@ func ValidateEndUserAuthenticationPolicySpecBinding(msg proto.Message) error {
 
 // ValidateVirtualService checks that a v1alpha3 route rule is well-formed.
 func ValidateVirtualService(msg proto.Message) (errs error) {
-	routeRule, ok := msg.(*networking.VirtualService)
+	virtualService, ok := msg.(*networking.VirtualService)
 	if !ok {
-		return errors.New("cannot cast to v1alpha3 routing rule")
+		return errors.New("cannot cast to virtual service")
 	}
 
-	// TODO: routeRule.Gateways
+	for _, gateway := range virtualService.Gateways {
+		if !IsDNS1123Label(gateway) { // FIXME: verify this restriction
+			errs = appendErrors(errs, fmt.Errorf("gateway is not a valid DNS1123 label: %v", gateway))
+		}
+	}
 
-	if len(routeRule.Hosts) == 0 {
-		errs = appendErrors(errs, fmt.Errorf("route rule must have at least one host"))
+	if len(virtualService.Hosts) == 0 {
+		errs = appendErrors(errs, fmt.Errorf("virtual service must have at least one host"))
 	}
-	for _, host := range routeRule.Hosts {
+	for _, host := range virtualService.Hosts {
 		errs = appendErrors(errs, validateHost(host))
 	}
 
-	for _, httpRoute := range routeRule.Http {
-		errs = appendErrors(validateHTTPRoute(httpRoute))
+	if len(virtualService.Http) == 0 && len(virtualService.Tcp) == 0 {
+		errs = appendErrors(errs, fmt.Errorf("http or tcp must be provided in virtual service"))
+	}
+	for _, httpRoute := range virtualService.Http {
+		errs = appendErrors(errs, validateHTTPRoute(httpRoute))
 	}
 
 	// TODO: validate once implemented
-	if len(routeRule.Tcp) > 0 {
-		errs = appendErrors(errs, errors.New("TCP route rules have not been implemented"))
+	if len(virtualService.Tcp) > 0 {
+		errs = appendErrors(errs, errors.New("TCP virtual services have not been implemented"))
```

Comment:
Oh ug, I didn't know this existed (I guess its a good thing we were eating errors from a TCP perspective). I implemented this last week; can you add validation for TCP matching? (If not I can in another PR.)

Reply:
I'll remove this check for now, and we can add the validation in another PR.